### ISSUE TITLE
Auto-start the jupyter notebook server if none exists

### ISSUE THREAD

**Issue by FelixKratz** [kiyoon/jupynium.nvim#12]
I think it would be really cool if a `jupyter notebook` session could be started from within nvim, e.g.

* I open a `*.ju.py` file in neovim
*  If there is no jupyer notebook server on 8888, jupynium starts a server for me and creates a self generated token, such that I don't have to enter it.
* Selenium starts and automatically accesses the jupyter notebook server with the correct token
*  The corresponding ipynb is opened

**Comment by kiyoon** [comment]
Thanks for the cool suggestion!

I think Jupyter Notebook doesn't provide me with the token with the API. I can launch the server but can't know the token. Maybe there's a way if I delve into their implementation..  
A little hacky way is I can maybe infer the URL from the output and open it. I'll look into this.  
In the meantime, you can use password rather than token. You can even diable the password completely, which will make it easier to connect.

**Comment by FelixKratz** [comment]
You can supply a custom token for the server to use as an additional argument along with a bunch of other things:
```bash
jupyter notebook --NotebookApp.token=1234
```
The directory can be specified via the argument `--NotebookApp.notebook_dir=dir`

**Comment by FelixKratz** [comment]
> @FelixKratz I originally thought about reading the URL but this looks simpler, although it will add some security risk. I think I will just make the command configurable and give you a default command template. I think the token has to be configurable also, but if it stays in the `init.lua` it is also a little risky

If the token is generated somewhat randomly it should be fine, as it is only hosted on the loopback interface anyways.

**Comment by FelixKratz** [comment]
If I have a *.ju.py file open I could imagine that it would be nice to have a command that starts the notebook server, starts jupynium and starts syncing the file to the server, such that it is an all-in-one command. When the notebook server is run from jupynium directly there is no reason to wait with syncing, because no manual input is needed. Maybe a new command? Something like: JupyniumStartNotebookServerFromBuffer or something a bit shorter ^^

**Comment by FelixKratz** [comment]
> If I have a *.ju.py file open I could imagine that it would be nice to have a command that starts the notebook server, starts jupynium and starts syncing the file to the server, such that it is an all-in-one command. When the notebook server is run from jupynium directly there is no reason to wait with syncing, because no manual input is needed. Maybe a new command? Something like: JupyniumStartNotebookServerFromBuffer or something a bit shorter ^^

If a server is already running on 8888 this command should probably fall back to attaching to the server and syncing the new file to a new ipynb instance

**Comment by kiyoon** [comment]
I wanted to make commands as self-explanatory as possible so users have no confusion. Do you fuzzy-find the commands? I'm using `wilder.nvim` and I start server by typing `:jara<tab>` for example. I get that it sometimes gets too long though.  
I can leave the current behaviour as is so that you can customise. You can make a custom function that calls both commands (or lua functions).  
And maybe I can add the fallback behaviour when server is not available, it will open the server and open the notebook if it exists and start syncing.

**Comment by FelixKratz** [comment]
> I wanted to make commands as self-explanatory as possible so users have no confusion. Do you fuzzy-find the commands? I'm using `wilder.nvim` and I start server by typing `:jara<tab>` for example. I get that it sometimes gets too long though.
> I can leave the current behaviour as is so that you can customise. You can make a custom function that calls both commands (or lua functions).
> And maybe I can add the fallback behaviour when server is not available, it will open the server and open the notebook if it exists and start syncing.

Nice, at some point I will create a keyboard shortcut to execute these commands.

**Comment by FelixKratz** [comment]
Another idea that I just had was that when the jupyter notebook process is started via jupynium in nvim it could be nice to have a popup window available in nvim that shows the console output of the jupyter notebook process. It is not important, but could be nice to have to check for the status of plugins etc.

**Comment by kiyoon** [comment]
By popup you mean another split or using something like plenary popup and make it look like Telescope? I've never used the popup feature before but in that case you can only see the output when it's open and you're not changing the buffer right?

**Comment by FelixKratz** [comment]
I had something like telescope in mind, but this is something I would maybe use only once or twice a month to be honest.

**Comment by kiyoon** [comment]
This has been almost implemented (except the popup stuff).

You can try the `feat/autostart-notebook` branch (`branch = 'feat/autostart-notebook'` on packer) to test it out before it gets released.

**Comment by kiyoon** [comment]
Now this is done. Thank you for the awesome feedback!

**Comment by FelixKratz** [comment]
First of all, thank you for adding this awesome feature! It is really nice.

I have noticed a problem however, it seems that the auto sync is not working, a tab appears and then instantly closes again:

https://user-images.githubusercontent.com/22680421/214052480-7fd90c29-3c5a-4c73-9009-278bdac2043a.mp4

This is my config: https://github.com/FelixKratz/dotfiles/blob/master/.config/nvim/lua/plugins/jupynium.lua

**Comment by kiyoon** [comment]
Oh, sorry for the trouble. I tested it on Mac also and it's working okay for me..  
So if you don't autostart the server and start syncing, it works?

**Comment by FelixKratz** [comment]
If I press 'n' it works

**Comment by FelixKratz** [comment]
> You mean if you press 'n', and then again run `:JupyniumStartSync` it works? I want to know if it's a problem caused by that function, or from the autostarting logic. I can't reproduce it now. Do you use Mac Homebrew python 3.10 or above? There's a bug that will be fixed from the pip side, which breaks the installation on mac. Maybe that's affecting this too (like maybe some files are not correctly installed)
> 
> Maybe can you try building with `pip3 install -e .` or `/usr/bin/pip3 install .`?
> 
> Or is it possible to run with CLI `nvim --listen /tmp/jupy_nvim` and `jupynium --nvim_listen_addr /tmp/jupy_nvim` and paste the output here?

No, if I press 'n' it works without doing anything else. So 'n' does what I thought 'y' would.

**Comment by kiyoon** [comment]
That's bizarre. It doesn't happen to me. 'n' opens a file and start syncing as well? I need to investigate why that happens

**Comment by FelixKratz** [comment]
> That's bizarre. It doesn't happen to me. 'n' opens a file and start syncing as well? I need to investigate why that happens

Yes, here is a video of that:

https://user-images.githubusercontent.com/22680421/214116407-de72042b-dda0-4221-9fc1-f7b732499356.mp4

**Comment by kiyoon** [comment]
That's really funny..

https://github.com/kiyoon/jupynium.nvim/blob/3b86f69a6893a475b44f24652b5a8fc63f3ae420/src/jupynium/cmds/jupynium.py#L403-L410

As you can see it should only do something when the input is `y`. Also can't really reproduce the hanging issue. What's your python version (I'm assuming Homebrew python 3.10)? I'll test on mac again and will come back to you.

**Comment by FelixKratz** [comment]
python version is
```sh
 # python --version
Python 3.9.7
```

**Comment by kiyoon** [comment]
Is that the one that installed Jupynium?

Will this work?
```bash
python -m jupynium --version
```

**Comment by FelixKratz** [comment]
> Is that the one that installed Jupynium?
> 
> Will this work?
> 
> ```shell
> python -m jupynium --version
> ```

```sh
 # python -m jupynium --version
Jupynium v0.1.dev20+g3b86f69
```

**Comment by FelixKratz** [comment]
Maybe there is something strange going on in my neovim....

**Comment by kiyoon** [comment]
Version looks fine. Mine is `Jupynium v0.1.1.dev41+g3b86f69` (I didn't know the formatting of version is different on different system) but the commit is at least the same

**Comment by kiyoon** [comment]
I finally reproduced. I'll see and let you know. Sorry for the hassle

**Comment by kiyoon** [comment]
The issue is when `auto_start_sync` is enabled it will start syncing on top of the fallback sync behaviour together. That's why `n` will sync normally as well and `y` clashes

**Comment by FelixKratz** [comment]
Ah yes, that makes sense!

**Comment by kiyoon** [comment]
It should be fixed now by #25
